velox/functions/prestosql/aggregates/PrestoHasher.cpp

```diff
@@ -0,0 +1,285 @@
+  return vector.isNullAt(row) ? 0 : hashInteger(vector.valueAt<T>(row));
+}
+
+FOLLY_ALWAYS_INLINE int64_t
+hashStringView(const DecodedVector& vector, vector_size_t row) {
+  if (vector.isNullAt(row)) {
+    return 0;
+  }
+  auto input = vector.valueAt<StringView>(row);
+  return XXH64(input.data(), input.size(), 0);
+}
+
+template <typename Callable>
+FOLLY_ALWAYS_INLINE void applyHashFunction(
+    const SelectivityVector& rows,
+    BufferPtr& hashes,
+    Callable func) {
+  VELOX_CHECK_GE(hashes->size(), rows.end())
+  auto rawHashes = hashes->asMutable<int64_t>();
+
+  rows.applyToSelected([&](auto row) { rawHashes[row] = func(row); });
+}
+
+template <typename T>
+FOLLY_ALWAYS_INLINE void hashIntegral(
+    const DecodedVector& vector,
+    const SelectivityVector& rows,
+    BufferPtr& hashes) {
+  applyHashFunction(
+      rows, hashes, [&](auto row) { return hashInteger<T>(vector, row); });
+}
+
+} // namespace
+
+template <TypeKind kind>
+FOLLY_ALWAYS_INLINE void PrestoHasher::hash(
+    const SelectivityVector& rows,
+    BufferPtr& hashes) {
+  using T = typename TypeTraits<kind>::NativeType;
+  hashIntegral<T>(*vector_.get(), rows, hashes);
+}
+
+template <>
+FOLLY_ALWAYS_INLINE void PrestoHasher::hash<TypeKind::BOOLEAN>(
+    const SelectivityVector& rows,
+    BufferPtr& hashes) {
+  applyHashFunction(rows, hashes, [&](auto row) {
+    return vector_->isNullAt(row)     ? 0
+        : vector_->valueAt<bool>(row) ? 1231
+                                      : 1237;
+  });
+}
+
+template <>
+FOLLY_ALWAYS_INLINE void PrestoHasher::hash<TypeKind::DATE>(
+    const SelectivityVector& rows,
+    BufferPtr& hashes) {
+  applyHashFunction(rows, hashes, [&](auto row) {
+    return vector_->isNullAt(row)
+        ? 0
+        : hashInteger(vector_->valueAt<Date>(row).days());
+  });
+}
+
+template <>
+FOLLY_ALWAYS_INLINE void PrestoHasher::hash<TypeKind::REAL>(
+    const SelectivityVector& rows,
+    BufferPtr& hashes) {
+  hashIntegral<int32_t>(*vector_.get(), rows, hashes);
+}
+
+template <>
+FOLLY_ALWAYS_INLINE void PrestoHasher::hash<TypeKind::VARCHAR>(
+    const SelectivityVector& rows,
+    BufferPtr& hashes) {
+  applyHashFunction(rows, hashes, [&](auto row) {
+    return hashStringView(*vector_.get(), row);
+  });
+}
+
+template <>
+FOLLY_ALWAYS_INLINE void PrestoHasher::hash<TypeKind::VARBINARY>(
+    const SelectivityVector& rows,
+    BufferPtr& hashes) {
+  applyHashFunction(rows, hashes, [&](auto row) {
+    return hashStringView(*vector_.get(), row);
+  });
+}
+
+template <>
+FOLLY_ALWAYS_INLINE void PrestoHasher::hash<TypeKind::DOUBLE>(
+    const SelectivityVector& rows,
+    BufferPtr& hashes) {
+  hashIntegral<int64_t>(*vector_.get(), rows, hashes);
+}
+
+template <>
+FOLLY_ALWAYS_INLINE void PrestoHasher::hash<TypeKind::TIMESTAMP>(
+    const SelectivityVector& rows,
+    BufferPtr& hashes) {
+  applyHashFunction(rows, hashes, [&](auto row) {
+    return vector_->isNullAt(row)
```

Comment:
Null handling is the same for all types. Consider moving it into applyHashFunction so that it is done once. This will also help ensure that null handling is done properly for all types.